src/Servers/Kestrel/Core/src/Internal/Http/Http1MessageBody.cs

```diff
@@ -159,6 +160,25 @@ protected async Task OnConsumeAsyncAwaited()
                 KestrelBadHttpRequestException.Throw(RequestRejectionReason.FinalTransferCodingNotChunked, transferEncoding);
             }
 
+            // https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2
+            // A sender MUST NOT send a Content-Length header field in any message
+            // that contains a Transfer-Encoding header field.
+            // https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.3
+            // If a message is received with both a Transfer-Encoding and a
+            // Content-Length header field, the Transfer-Encoding overrides the
+            // Content-Length.  Such a message might indicate an attempt to
+            // perform request smuggling (Section 9.5) or response splitting
+            // (Section 9.4) and ought to be handled as an error.  A sender MUST
+            // remove the received Content-Length field prior to forwarding such
+            // a message downstream.
+            // We should remove the Content-Length request header in this case, for compatibility
+            // reasons, increase x-Content-Length so that the original Content-Length is still available.
+            if (headers.ContentLength.HasValue)
+            {
+                context.RequestHeaders.Add($"x-{HeaderNames.ContentLength}", context.RequestHeaders[HeaderNames.ContentLength]);
+                context.RequestHeaders.Remove(HeaderNames.ContentLength);
```

Comment:
Using the property is more efficient. Also, we don't want to re-create this string each time.
```suggestion
                headers.Add("X-Content-Length", headers[HeaderNames.ContentLength]);
                headers.ContentLength = null;
```

Reply:
Here `headers.Add("X-Content-Length", headers[HeaderNames.ContentLength]);`, `headers[HeaderNames.ContentLength]` should be changed to use `headers.ContentLength`?

Reply:
I avoided that because headers.ContentLength is a long where the indexer returns a StringValues, and we need a StringValues.